extensions/yarn/main.ts

```diff
@@ -85,9 +92,57 @@ if (Deno.args.length >= 1
         || Deno.args[0] === 'dedupe'
    ) {
     await checkDryRun(Deno.args[0], Deno.args.slice(1));
-}
 
-// Run the command with side effects.
-console.log(`[${green("phylum")}] Applying changes…`);
-let status = await Deno.run({ cmd: ['yarn', ...Deno.args] }).status();
-Deno.exit(status.code);
+    console.log(`[${green("phylum")}] Downloading packages to cache…`);
+
+    // Download packages to cache without sandbox.
+    let status = await Deno.run({ cmd: ['yarn', ...Deno.args, '--mode=skip-build']}).status();
+
+    // Ensure download worked. Failure is still "safe" for the user.
+    if (!status.success) {
+        console.error(`[${red("phylum")}] Downloading packages to cache failed.\n`);
+        Deno.exit(status.code);
+    } else {
+        console.log(`[${green("phylum")}] Cache updated successfully.\n`);
+    }
+
+    console.log(`[${green("phylum")}] Building packages inside sandbox…`);
+
+    // Run build inside a sandbox.
+    const output = PhylumApi.runSandboxed({
+        cmd: 'yarn',
+        args: ['install', '--immutable', '--immutable-cache'],
+        exceptions: {
+            run: ['yarn', 'node'],
+            read: ['./', '~/.cache/node/corepack'],
+            write: ['./.yarn/install-state.gz'],
+            net: false,
+        },
+    });
+
+    // Failure here could indicate vulnerabilities; report to the user.
+    if (!output.success) {
+        console.log(`[${red("phylum")}] Sandboxed build failed.`);
+        console.log(`[${red("phylum")}]`);
+
+        // Check for SIGSYS, since this confirms sandbox violation.
+        if (output.signal === 31) {
+            console.log(`[${red("phylum")}] Sandbox violation confirmed during install.`);
+            console.log(`[${red("phylum")}] Do NOT retry installation.`);
+            console.log(`[${red("phylum")}]`);
+            console.log(`[${red("phylum")}] Please submit your lockfile to Phylum.`);
+        } else {
+            console.log(`[${red("phylum")}] This could mean one of your packages attempted to access a restricted resource.`);
+            console.log(`[${red("phylum")}] Do not retry installation without Phylum's extension.`);
+            console.log(`[${red("phylum")}]`);
+            console.log(`[${red("phylum")}] Please submit your lockfile to Phylum should this error persist.`);
+        }
```

Comment:
Feedback on error messages would be appreciated, we'll probably want to share this wording across our extensions and I'm not sure this is ideal already.

Reply:
The error message is pretty good as is. I've wondered if we should say anything about the possible error message that `yarn` may have output above this message, but I'm really not sure what to say there.

We could probably include a link with the "Please submit" wording. Something like this:

```
Please consider reporting this to Phylum using the following link:
https://github.com/phylum-dev/cli/issues/new?template=sandbox_violation.md
```

We would need to add `.github/ISSUE_TEMPLATE/sandbox_violation.md` for that link to work, but we could add prompts in that template for details such as full command output and the lockfile.

Reply:
Cc @furi0us333 since he may have opinions on the wording.